File: p2644.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.StringTokenizer;

public class p2644 {
    public static ArrayList<ArrayList<Integer>> family;
    public static boolean[] visited;
    public static int res;

    public static void dfs(int src, int dest, int cnt) {
        for (int i = 0; i < family.get(src).size(); i++) {
            if (!visited[family.get(src).get(i)]) {
                if (dest == family.get(src).get(i)) {
                    res = cnt + 1;
                    return;
                }
                visited[family.get(src).get(i)] = true;
                dfs(family.get(src).get(i), dest, cnt + 1);
            }
        }
        return;
    }

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        String srcDest = br.readLine();
        int src = Integer.parseInt(srcDest.split(" ")[0]);
        int dest = Integer.parseInt(srcDest.split(" ")[1]);
        int m = Integer.parseInt(br.readLine());
        visited = new boolean[n + 1];
        family = new ArrayList<>();

        for (int i = 0; i <= n; i++) {
            family.add(new ArrayList<>());
        }

        for (int i = 0; i < m; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            int x = Integer.parseInt(st.nextToken());
            int y = Integer.parseInt(st.nextToken());

            family.get(x).add(y);
            family.get(y).add(x);
        }
        res = -1;
        dfs(src, dest, 0);
        System.out.println(res);
    }
}

File: p2644.cpp
#include<bits/stdc++.h>
using namespace std;

vector<bool> visited;
int answer = -1;

void dfs(vector<vector<int>> map, int src, int dest, int depth){
    
    for(int i = 0; i < map[src].size(); i++){
        if(visited[map[src][i]]) continue;
        if(map[src][i] == dest){
            answer = depth + 1;
            return;
        }

        visited[map[src][i]] = true;
        dfs(map, map[src][i], dest, depth + 1);
    }
}

int main(){
    int N;
    cin >> N;
    vector<vector<int>> map(N + 1, vector<int>());

    for(int i = 0; i < N + 1; i++) visited.push_back(false);

    int src, dest, r;

    cin >> src >> dest;
    cin >> r;

    for(int i = 0; i < r; i++){
        int x, y;
        cin >> x >> y;
        map[x].push_back(y);
        map[y].push_back(x);
    }

    dfs(map, src, dest, 0);
    
    cout << answer << "\n";
}